### DISCUSSION TITLE
Is there any built in @defer support?

### DISCUSSION THREAD

**Discussion by Roundaround** [graphql-java-kickstart/graphql-java-tools#324] [Q&A] [answered]
I've been racking my brain trying to figure out how to use graphql-java-tools for the resolvers and autogenerated GraphQL endpoint together with graphql-java's support for the defer directive. I already have a BatchLoader returning the values via CompletableFuture.supplyAsync so it feels like it _should_ be an easy wire-up.

Am I missing something?

**Comment by neerajsu** [comment]
I have the same question (#433 ). Wish one of the authors provided some example, code snippet or documentation.

**Comment by neerajsu** [comment, marked as answer]
@Roundaround

Okay I figured out a small clean hack to get it to work with minimal effort

[graphql-java project](https://github.com/graphql-java/graphql-java)  has inbuilt `@defer` support. The directives can be defined while creating a GraphQLSchema Object. Previously `@defer` directive was enabled by default. But now it is **disabled** by default. So, you need to enable it by adding `graphql.Directives.DeferDirective` to the set of derectives.

In this graphql-java-tools project, you can technically create a GraphQLSchema object and expose it as a bean and the kickstart autoconfigure will pick it up. Problem is, when you create a GraphQLSchema object you need to define all the resolvers and types etc, which is all boilerplate code. This project automatically autowires all of it using spring. So doing it in the graphql-java way would defeat the purpose of using this project.

To circumvent that, for now, unless the authors of this project find a more elegant way to support this, you can get `@defer` to work using a small snippet of code. I'm basically getting the GraphQLSchema object's existing list of directives and adding graphQLDirectives to it using [FieldUtils](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/reflect/FieldUtils.html) (Reflection).

Add this bean in any @Configuration class

```java
    @Bean
    public GraphQLDirective deferDirective(GraphQLSchemaServletProvider graphQLSchemaServletProvider) throws IllegalAccessException {
        GraphQLSchema graphQLSchema = graphQLSchemaServletProvider.getSchema();
        //getting existing directives
        Set<GraphQLDirective> graphQLDirectives = new HashSet<>(graphQLSchema.getDirectives());
       //adding Defer directive to the existing set
        graphQLDirectives.add(Directives.DeferDirective);
        //setting private field using reflection since there is no public way to modify this variable
        FieldUtils.writeField(graphQLSchema, "directives", graphQLDirectives, true);
        return Directives.DeferDirective;
    }
```

And that's all, you now have enabled defer support. You don't need to modify any existing code, or return CompletableFuture, Publisher etc. Its all just done by the engine.

When you send a query to the server, you can just add @defer to any field in a query and that field's result will be deferred. 

Ex:-  In this query, postText resolves quickly whereas comments and reviews would show up later when they are computed resolved using service side events.
```graphql
query {
           post {
               postText
               comments @defer {
                   commentText
               }
               reviews @defer {
                   reviewText {
               }
        }
```

Note:- the type of the deferred field has to be nullable (So you cannot use !)
